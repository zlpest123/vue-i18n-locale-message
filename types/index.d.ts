import { SFCDescriptor } from 'vue-template-compiler'

/**
 *  Locale Message Recursive Structure
 *    e.g.
 *    {
 *      "en": {  // 'en' Locale
 *        "key1": "this is message1", // basic
 *        "nested": { // nested
 *          "message1": "this is nested message1"
 *        },
 *        "errors": [ // array
 *          "this is 0 error code message",
 *          {  // object in array
 *            "internal1": "this is internal 1 error message"
 *          },
 *          [  // array in array
 *            "this is nested array error 1"
 *          ]
 *        ]
 *      },
 *      "ja": { // 'ja' Locale
 *        // ...
 *      }
 *    }
 */

export type Locale = string
export type LocaleMessage = string | LocaleMessageObject | LocaleMessageArray
export interface LocaleMessageArray extends Array<LocaleMessage> {}
export interface LocaleMessageObject { [key: string]: LocaleMessage }
export type LocaleMessages = Record<Locale, LocaleMessageObject>

/**
 *  SFC (Single-file component) file info
 *    e.g.
 *    {
 *      path: '/path/to/project1/src/components/common/Modal.vue',
 *      content: `
 *        <template>
 *          <!-- template contents is here ... -->
 *        </template>
 *
 *        <script>
 *          // script codes is here ...
 *        </script>
 *
 *        <style scoped>
 *          // css style codes is here ...
 *        </style>
 *
 *        <i18n>
 *        {
 *          "en": {
 *            "ok": "OK",
 *            "cancel": "Cancel"
 *          },
 *          "ja": {
 *            "ok": "OK",
 *            "cancel": "キャンセル"
 *          }
 *        }
 *        </i18n>
 *      `,
 *    }
 */

export interface SFCFileInfo {
  path: string
  content: string
}

/**
 *  Extend SFCDescriptor, due to squeeze / infuse.
 *    e.g.
 *    {
 *      template: { ... },
 *      script: { ... },
 *      styles: [{ ... }],
 *      customBlocks: [{
 *        type: 'i18n',
 *        content: `
 *        {
 *          "en": {
 *            "ok": "OK",
 *            "cancel": "Cancel"
 *          },
 *          "ja": {
 *            "ok": "OK",
 *            "cancel": "キャンセル"
 *          }
 *        }
 *        `,
 *        attrs: { ... },
 *        start: 10,
 *        end: 30,
 *        map: { ... }
 *      }, ...],
 *      contentPath: '/path/to/project1/src/components/common/Modal.vue',
 *      component: 'Modal',
 *      hierarchy: ['components', 'common', 'Modal']
 *    }
 */

// extend for vue-i18n-locale-message
declare module 'vue-template-compiler' {
  interface SFCDescriptor {
    raw: string
    contentPath: string
    component: string
    hierarchy: string[]
  }
}
