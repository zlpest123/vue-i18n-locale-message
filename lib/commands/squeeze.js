"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
const squeezer_1 = __importDefault(require("../squeezer"));
const fs_1 = __importDefault(require("fs"));
exports.command = 'squeeze';
exports.aliases = 'sqz';
exports.describe = 'squeeze locale messages from single-file components';
exports.builder = (args) => {
    const outputDefault = `${process.cwd()}/messages.json`;
    return args
        .option('target', {
        type: 'string',
        alias: 't',
        describe: 'target path that single-file components is stored',
        demandOption: true
    })
        .option('output', {
        type: 'string',
        alias: 'o',
        default: outputDefault,
        describe: 'path to output squeezed locale messages'
    });
};
exports.handler = (args) => {
    const targetPath = utils_1.resolve(args.target);
    const messages = squeezer_1.default(targetPath, utils_1.readSFC(targetPath));
    writeLocaleMessages(utils_1.resolve(args.output), messages);
};
function writeLocaleMessages(output, messages) {
    // TODO: async implementation
    fs_1.default.writeFileSync(output, JSON.stringify(messages, null, 2));
}
exports.default = {
    command: exports.command,
    aliases: exports.aliases,
    describe: exports.describe,
    builder: exports.builder,
    handler: exports.handler
};
