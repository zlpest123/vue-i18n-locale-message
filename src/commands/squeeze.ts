import { Arguments, Argv } from 'yargs'
import { LocaleMessages } from '../../types'

import { resolve, readSFC } from '../utils'
import squeeze from '../squeezer'
import fs from 'fs'

type SqueezeOptions = {
  target: string
  output: string
}

export const command = 'squeeze'
export const aliases = 'sqz'
export const describe = 'squeeze locale messages from single-file components'

export const builder = (args: Argv): Argv<SqueezeOptions> => {
  const outputDefault = `${process.cwd()}/messages.json`
  return args
    .option('target', {
      type: 'string',
      alias: 't',
      describe: 'target path that single-file components is stored',
      demandOption: true
    })
    .option('output', {
      type: 'string',
      alias: 'o',
      default: outputDefault,
      describe: 'path to output squeezed locale messages'
    })
}

export const handler = (args: Arguments<SqueezeOptions>): void => {
  const targetPath = resolve(args.target)
  const messages = squeeze(targetPath, readSFC(targetPath))
  writeLocaleMessages(resolve(args.output), messages)
}

function writeLocaleMessages (output: string, messages: LocaleMessages) {
  // TODO: async implementation
  fs.writeFileSync(output, JSON.stringify(messages, null, 2))
}

export default {
  command,
  aliases,
  describe,
  builder,
  handler
}
