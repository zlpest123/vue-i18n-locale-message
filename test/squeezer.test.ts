import squeeze from '../src/squeezer'
import metaInfo from './fixtures/meta'
import yamlMetaInfo from './fixtures/format/yaml'
import json5MetaInfo from './fixtures/format/json5'
import { LocaleMessages, LocaleMessageMeta } from '../types'

test('basic', () => {
  const mesasges: LocaleMessages = squeeze(metaInfo as LocaleMessageMeta[])
  expect(mesasges).toMatchSnapshot()
})

test('yaml', () => {
  const mesasges: LocaleMessages = squeeze(yamlMetaInfo as LocaleMessageMeta[])
  expect(mesasges).toMatchSnapshot()
})

test('json5', () => {
  const mesasges: LocaleMessages = squeeze(json5MetaInfo as LocaleMessageMeta[])
  expect(mesasges).toMatchSnapshot()
})
