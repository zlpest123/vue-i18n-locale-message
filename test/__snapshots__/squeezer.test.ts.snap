// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic 1`] = `
Object {
  "en": Object {
    "App": Object {
      "title": "Application",
    },
    "components": Object {
      "Modal": Object {
        "cancel": "Cancel",
        "ok": "OK",
      },
      "nest": Object {
        "RankingTable": Object {
          "headers": Object {
            "name": "Name",
            "rank": "Rank",
            "score": "Score",
          },
        },
      },
    },
  },
  "ja": Object {
    "App": Object {
      "title": "アプリケーション",
    },
    "components": Object {
      "Modal": Object {
        "cancel": "キャンセル",
        "ok": "OK",
      },
    },
    "pages": Object {
      "Login": Object {
        "button": "ログイン",
        "confirm": "パスワードの確認入力",
        "id": "ユーザーID",
        "passowrd": "パスワード",
      },
    },
  },
}
`;

exports[`json5 1`] = `
Object {
  "en": Object {
    "components": Object {
      "Modal": Object {
        "cancel": "Cancel",
        "ok": "OK",
      },
    },
  },
  "ja": Object {
    "components": Object {
      "Modal": Object {
        "cancel": "キャンセル",
        "ok": "OK",
      },
    },
  },
}
`;

exports[`yaml 1`] = `
Object {
  "en": Object {
    "components": Object {
      "Modal": Object {
        "cancel": "Cancel",
        "ok": "OK",
      },
    },
  },
  "ja": Object {
    "components": Object {
      "Modal": Object {
        "cancel": "キャンセル",
        "ok": "OK",
      },
    },
  },
}
`;
